### PR TITLE
AVX2 transcendentals intrinsics

### PR DESCRIPTION
**Pull request by vkarak**
Provides the following transcendentals intrinsics:

* `nmc_mm256_exp_pd`
* `nmc_mm256_log_pd`
* `nmc_mm256_pow_pd`
* `nmc_mm256_frexp_pd`

The first three are the equivalent of the corresponding SVML intrinsics without the `nmc` prefix.

The last one is used by the `log` function, but I decided it's nice to have it public.

All results are tested against full precision standard library implementation and provide equal results (except for the `pow` case). `NaN`s and infinities are treated according to the standard.

Limitations:

* Subnormals are treated as zeros by `frexp` and as a result by `log`.
* The `pow` intrinsic is expressed as function of the `log` and `exp`. Therefore it is not very efficient and also introduces higher errors compared to the standard library's `pow` implementation.

Performance against SVML:

I have not done extensive performance testing against the SVML's implementation (I believe we might be a bit slower though), but in the context of the miniapp, GCC now emits equally fast code as the Intel compiler with SVML.

Other:

Fixed GCC warnings for the auto generated code. The fix isn't sth smart; just a bunch of `unused` attributes.

### CODE REVIEW

File: modcc/backends/avx2.hpp [hunk 1]
```diff
@@ -23,8 +23,12 @@ struct simd_intrinsics<targetKind::avx2> {
     }
 
     static std::string emit_headers() {
-        return "#include <immintrin.h>";
-    };
+        std::string ret = "#include <immintrin.h>";
+#ifndef __INTEL_COMPILER
```
[the remark this answers is not in the document]
Comment (vkarak) [reply modcc/backends/avx2.hpp]: Sure.

File: src/backends/multicore/intrin.hpp [hunk 2]
```diff
@@ -0,0 +1,312 @@
+//
+// Custom transcendental intrinsics
+//
+// Implementation inspired by the VDT and Cephes libraries:
+//    - https://github.com/dpiparo/vdt
+//    - http://www.netlib.org/cephes/
```
[the remark this answers is not in the document]
Comment (vkarak) [reply src/backends/multicore/intrin.hpp]: My implementation was initially based on VDT, which is in turn is based on Cephes. Cephes has no license associated with it. Besides the fully vectorised low-level implementation with intrinsics, another key difference between my implementation and that of VDT is the `frexp()` function that is used by the `log()` implementation. VDT goes with a custom implementation, whereas mine follows the look-and-feel of the rest of the intrinsics and adheres to the standard both from the API's point of view and the behaviour.

Of course I can expand more on the "inspiration" part in the comments.

A common thing about all these libraries are using Padé approximation for computing these transcendentals, so we can base our comments on that. I would perhaps need your math knowledge here.

File: src/backends/multicore/intrin.hpp [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+#include <iostream>
+#include <immintrin.h>
+
+namespace nest {
+namespace mc {
+namespace multicore {
+
+namespace detail {
+
+const double exp_limit = 708;
```
[the remark this answers is not in the document]
Comment (vkarak) [reply src/backends/multicore/intrin.hpp]: Sure they can.

File: src/backends/multicore/intrin.hpp [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+const double qx1exp = 3.00198505138664455042E-6;
+const double qx2exp = 2.52448340349684104192E-3;
+const double qx3exp = 2.27265548208155028766E-1;
+const double qx4exp = 2.00000000000000000009E0;
+const double qx1log = 1.12873587189167450590e1;
+const double qx2log = 4.52279145837532221105e1;
+const double qx3log = 8.29875266912776603211e1;
+const double qx4log = 7.11544750618563894466e1;
+const double qx5log = 2.31251620126765340583E1;
+const double log2e  = 1.4426950408889634073599; // 1/log(2)
```
[the remark this answers is not in the document]
Comment (vkarak) [reply src/backends/multicore/intrin.hpp]: Indeed. However, their names are based on the P/Q polynomials used in the method and some obvious constants, but I could some of our salt.

File: tests/unit/test_intrin.cpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
+#include <cmath>
+#include <limits>
+#include <backends/multicore/intrin.hpp>
+#include <immintrin.h>
+
+#include "../gtest.h"
+//#include "../../../vdt/include/exp.h"
```
[the remark this answers is not in the document]
Comment (vkarak) [reply tests/unit/test_intrin.cpp]: Yep.

File: modcc/backends/avx2.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (vkarak) [reply modcc/backends/avx2.hpp]: Would it make sense to put this in `util/compat.hpp`, making it also a bit more generic, taking version numbers as well?

[the remark this answers is not in the document]
Comment (bcumming) [reply modcc/backends/avx2.hpp]: @halfflat, what do you think about @vkarak's proposal?

File: src/backends/multicore/intrin.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bcumming) [reply src/backends/multicore/intrin.hpp]: I think that this certainly needs to clarified.

If we include some documentation of the mathematical methods, or a link to some document that describes the methods, and add some comments to the code itself, I think that the variable name concerns below will be cleared up.

File: ATTRIBUTIONS.md [hunk 6]
```diff
@@ -29,3 +29,10 @@ project.
 BSD License.
 
 http://llvm.org/
+
+## Transcendentals intrinsics
+
+The numerical algorithms for the transcendentals intrinsics are based on the scalar Cephes library.
+
+http://www.netlib.org/cephes/
+No License.
```
Comment (vkarak) [inline ATTRIBUTIONS.md]: Should we contact here the Cephes developers for granting us permission to use their algorithm? Some libraries do that and state it explicitly, some not.

File: src/util/compat.hpp [hunk 7]
```diff
@@ -7,6 +7,16 @@
 
 namespace compat {
 
+constexpr bool using_intel_compiler(int major=0, int minor=0, int patchlevel=0) {
```
Comment (bcumming) [inline src/util/compat.hpp]: I like this interface.

File: src/backends/multicore/intrin.hpp [hunk 8]
```diff
@@ -93,42 +103,101 @@ void nmc_mm256_print_epi64x(__m256i x, const char *name) {
     std::cout << "}\n";
 }
 
+//
+// Calculates exponential using AVX2 instructions
```
Comment (bcumming) [inline src/backends/multicore/intrin.hpp]: This documentation, along with the inline comments, is great!

File: modcc/backends/avx2.hpp [hunk 9]
```diff
@@ -84,7 +98,11 @@ struct simd_intrinsics<targetKind::avx2> {
 
     template<typename B, typename E>
     static void emit_pow(TextBuffer& tb, const B& base, const E& exp) {
+#ifdef __INTEL_COMPILER
```
Comment (vkarak) [inline modcc/backends/avx2.hpp]: I should user the new compiler check in `compat` here.

**Pull request merged**